baiji/connection.py

```diff
@@ -0,0 +1,700 @@
+        '''
+        Acts like open(key, mode), opening a file.
+
+        If the file is on S3, it is downloaded and a NamedTemporaryFile is returned.
+
+        The caller is responsible for closing the file descriptor, please try to
+        use it in a with block.
+
+        Like os.open, after stripping 'U' from the mode, the first character must
+        be 'r' or 'w'. open also accepts 'x' for exclusive creation: it raises an
+        exception if the file already exists. This is meant as a convenience and
+        sanity check when you know the original file isn't supposed to exist.
+
+        Note that 'x' is not suitable for preventing concurrent creation. This is
+        due to the "eventually consistent" nature of s3, and also the CachedFile
+        design. If two processes simultaneously write to the same file, there's a
+        race condition. One or both will appear to succeed, and one will win out.
+        Using 'x' doesn't -- can't -- prevent this.
+
+        Raises s3.KeyNotFound when attempting to open a local or remote file that
+        doesn't exist, s3.KeyExists when attempting exclusive create on an
+        existing file or key, ValueError for an invalid key, and IOError for an
+        underlying local file-system failure.
+
+        '''
+        from baiji.cached_file import CachedFile
+        return CachedFile(key, mode, connection=self)
+
+    def list_buckets(self):
+        '''
+        List all the buckets availiable on AWS S3 for the credentialed account.
+        '''
+        return [x.name for x in self.conn.get_all_buckets()]
+
+    def bucket_info(self, name):
+        '''
+        Get info about a bucket
+        '''
+        bucket = self._bucket(name)
+        def safe_get(method):
+            from boto import exception
+            try:
+                return getattr(bucket, method)()
+            except exception.S3ResponseError:
+                return None
+        versioning_status = safe_get('get_versioning_status')
+        if versioning_status is not None and 'Versioning' in versioning_status:
+            versioning_status = versioning_status['Versioning']
+        else:
+            versioning_status = "Not Versioned"
+        return {
+            'name': name,
+            'lifecycle': safe_get('get_lifecycle_config'),
+            'cors': safe_get('get_cors'),
+            'location': safe_get('get_location'),
+            'logging_status': safe_get('get_logging_status'),
+            'policy': safe_get('get_policy'),
+            'versioning_status': versioning_status,
+            'website_configuration': safe_get('get_website_configuration'),
+            'website_endpoint': safe_get('get_website_endpoint'),
+        }
+
+    def create_bucket(self, name, versioned=True):
+        bucket = self.conn.create_bucket(name)
+        if versioned:
+            self.configure_versioning(name, True)
+        return bucket
+
+    def configure_versioning(self, name, state=True):
```

Comment:
Is `"state"` used in the AWS documentation? Something like `versioned=True` might be more clear.